covsirphy/cleaning/jhu_complement.py

```diff
@@ -197,10 +201,53 @@ def _monotonic(self, df, variable):
             series.fillna(method="ffill", inplace=True)
             # Reduce values to the previous date
             df.loc[:date, variable] = series * raw_last / series.iloc[-1]
-            df[variable] = df[variable].fillna(0).astype(np.int64)
+            df[variable] = df[variable].fillna(0)
+        df[variable] = df[variable].apply(np.ceil).astype(np.int64)
         self.complement_dict[f"Monotonic_{variable.lower()}"] = True
         return df
 
+    def _validate_recovery_period(self, df, upper_limit_days=90, lower_limit_days=7,
+                                  upper_percentage=0.5, lower_percentage=0.5):
+        """
+        Calculates and validates recovery period for specific country
+        as an additional condition in order to apply full complement or not
+
+        Args:
+            df (pandas.DataFrame):
+                - Index: reset_index
+                - Columns: Date, Confirmed, Recovered, Fatal
+            upper_limit_days (int): maximum number of valid partial recovery periods [days]
+            lower_limit_days (int): minimum number of valid partial recovery periods [days]
+            upper_percentage (float): fraction of partial recovery periods with value greater than upper_limit_days
+            lower_percentage (float): fraction of partial recovery periods with value less than lower_limit_days
+
+        Returns:
+            bool: true if recovery period is within valid range or false otherwise
+
+        Note: Passed argument df corresponds to specific country's df
+              upper_limit_days has default value of 3 months (90 days)
+              lower_limit_days has default value of 1 week (7 days)
+        """
+        recovery_period_is_valid = True
+        check_df = df.copy()
+        # Calculate "Confirmed - Fatal"
+        check_df["diff"] = check_df[self.C] - check_df[self.F]
+        check_df = check_df.loc[:, ["diff", self.R]]
+        # Calculate how many days passed to reach the number of cases
+        check_df = check_df.unstack().reset_index()
+        check_df.columns = ["Variable", "Date", "Number"]
+        check_df["Days"] = (check_df[self.DATE] - check_df[self.DATE].min()).dt.days
+        # Calculate partial recovery periods
+        check_df = check_df.pivot_table(values="Days", index="Number", columns="Variable")
+        check_df = check_df.interpolate(limit_area="inside").dropna().astype(np.int64)
+        check_df["Elapsed"] = check_df[self.R] - check_df["diff"]
+        check_df = check_df.loc[check_df["Elapsed"] > 0]
+        # Check partial recovery periods
+        if sum(check_df["Elapsed"] > upper_limit_days)/len(check_df) > upper_percentage or \
```

Comment:
Flake8 test failed.
Please replace
```Python
if sum(check_df["Elapsed"] > upper_limit_days)/len(check_df) > upper_percentage or \
           sum(check_df["Elapsed"] < lower_limit_days)/len(check_df) > lower_percentage:
            recovery_period_is_valid = False
        return recovery_period_is_valid
```
with 
```Python
per_up = (check_df["Elapsed"] > self.upper_limit_days).sum() / len(check_df)
per_lw = (check_df["Elapsed"] < self.lower_limit_days).sum() / len(check_df)
return per_up < self.upper_percentage and per_lw < self.lower_percentage
```
This change enable us to remove line 231 `recovery_period_is_valid = True`.

Reply:
We may spare the True value line but using this inverse logic is not very representative of what we consider as valid values and could confuse the reader of our intent.
When the flake8 test fails?

Reply:
Flake8/Sourcery AI say `recovery_period_is_valid = False; return recovery_period_is_valid` is unnecessary and spaces should be before/after `/`.

Please give me the details why this logic (De Morgan's laws) is confusing.

`per_up < self.upper_percentage and per_lw < self.lower_percentage` calculate True/False at the same time.

Reply:
No I understand the inverse law, it is of course correct. I just wanted to give the intend that the condition is always valid, except for when we have many values over upper or under lower limit. In the inverse logic, it states that it is true if the "bad" values (either over upper or under lower) are only few and not many, which is not the first thing someone will think as comparisons. But the code is more neat I guess so I will make the suggested change.